### PR TITLE
Add eth_estimateDiffSize

### PR DESCRIPTION
**Pull request by kpp**
# Description

## Linked Issues
- Fixes https://github.com/chainwayxyz/citrea/issues/420

## Testing
`make test`

## Docs
Adds `eth_estimateDiffSize` endpoint

### CODE REVIEW

File: crates/evm/src/query.rs [hunk 1]
```diff
@@ -43,25 +44,41 @@ pub(crate) const MIN_TRANSACTION_GAS: u64 = 21_000u64;
...
+    diff_size: u64,
 }
 
-impl EstimatedGas {
-    /// Return total estimated gas including evm gas and L1 fee.
-    pub(crate) fn total(&self) -> U256 {
-        self.l1_fee_overhead + U256::from(self.gas)
+impl EstimatedTxExpenses {
+    /// Return total estimated gas used including evm gas and L1 fee.
+    pub(crate) fn gas_spent(&self) -> U256 {
```
Comment (eyusufatik) [inline crates/evm/src/query.rs:65]: we can call this method gas_with_l1_overhead imo

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: crates/evm/src/query.rs [hunk 2]
```diff
@@ -43,25 +44,41 @@ pub(crate) const MIN_TRANSACTION_GAS: u64 = 21_000u64;
 /// <https://github.com/ethereum/go-ethereum/blob/a5a4fa7032bb248f5a7c40f4e8df2b131c4186a4/internal/ethapi/api.go#L56>
 const ESTIMATE_GAS_ERROR_RATIO: f64 = 0.015;
 
-/// The result of gas estimation.
+/// The result of gas/diffsize estimation.
```
Comment (rakanalh) [inline crates/evm/src/query.rs:47]:
```suggestion
/// The result of gas/delta estimation.
```

File: crates/evm/src/query.rs [hunk 3]
```diff
@@ -43,25 +44,41 @@ pub(crate) const MIN_TRANSACTION_GAS: u64 = 21_000u64;
...
     /// Evm gas used.
-    gas: U64,
-    /// Actually not an L1 fee but l1_fee / base_fee.
-    l1_fee_overhead: U256,
+    gas_used: U64,
+    /// Base fee of the L2 block when tx was executed.
+    base_fee: U256,
+    /// L1 fee.
+    l1_fee: U256,
+    /// L1 diff size.
```
Comment (rakanalh) [inline crates/evm/src/query.rs:59]:
```suggestion
    /// L1 gas delta.
```

File: crates/evm/src/query.rs [hunk 4]
```diff
@@ -43,25 +44,41 @@ pub(crate) const MIN_TRANSACTION_GAS: u64 = 21_000u64;
...
+    diff_size: u64,
```
Comment (rakanalh) [inline crates/evm/src/query.rs:60]:
```suggestion
    delta: u64,
```

File: crates/evm/src/query.rs [hunk 5]
```diff
@@ -43,25 +44,41 @@ pub(crate) const MIN_TRANSACTION_GAS: u64 = 21_000u64;
...
+        // Actually not an L1 fee but l1_fee / base_fee.
+        let l1_fee_overhead = U256::from(1).max(self.l1_fee / self.base_fee);
+        l1_fee_overhead + U256::from(self.gas_used)
     }
 }
 
+/// Result of estimation of diff size.
+#[derive(Clone, Default, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct EstimatedDiffSize {
```
Comment (rakanalh) [inline crates/evm/src/query.rs:75]:
```suggestion
pub struct EstimatedGasDelta {
```

File: crates/evm/src/query.rs [hunk 6]
```diff
@@ -43,25 +44,41 @@ pub(crate) const MIN_TRANSACTION_GAS: u64 = 21_000u64;
...
+impl EstimatedTxExpenses {
+    /// Return total estimated gas used including evm gas and L1 fee.
+    pub(crate) fn gas_spent(&self) -> U256 {
+        // Actually not an L1 fee but l1_fee / base_fee.
+        let l1_fee_overhead = U256::from(1).max(self.l1_fee / self.base_fee);
+        l1_fee_overhead + U256::from(self.gas_used)
     }
 }
 
+/// Result of estimation of diff size.
```
Comment (rakanalh) [inline crates/evm/src/query.rs:72]:
```suggestion
/// Result of estimation of delta.
```

File: crates/evm/src/query.rs [hunk 7]
```diff
@@ -43,25 +44,41 @@ pub(crate) const MIN_TRANSACTION_GAS: u64 = 21_000u64;
...
+#[derive(Clone, Default, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct EstimatedDiffSize {
+    /// Gas used.
+    pub gas: U64,
+    /// Diff size.
```
Comment (rakanalh) [inline crates/evm/src/query.rs:78]:
```suggestion
    /// Detla.
```

File: crates/evm/src/query.rs [hunk 8]
```diff
@@ -43,25 +44,41 @@ pub(crate) const MIN_TRANSACTION_GAS: u64 = 21_000u64;
...
+    pub diff_size: U64,
```
Comment (rakanalh) [inline crates/evm/src/query.rs:79]:
```suggestion
    pub delta: U64,
```

File: crates/evm/src/query.rs [hunk 9]
```diff
@@ -868,15 +883,49 @@ impl<C: sov_modules_api::Context> Evm<C> {
...
+        working_set: &mut WorkingSet<C>,
+    ) -> RpcResult<reth_primitives::U256> {
+        info!("evm module: eth_estimateGas");
+
+        let estimated = self.estimate_tx_expenses(request, block_number, working_set)?;
+        Ok(estimated.gas_spent())
+    }
+
+    /// Handler for: `eth_estimateDiffSize`
+    #[rpc_method(name = "eth_estimateDiffSize")]
```
Comment (rakanalh) [inline crates/evm/src/query.rs:912]:
```suggestion
    #[rpc_method(name = "eth_estimateGasDelta")]
```

File: crates/evm/src/query.rs [hunk 10]
```diff
@@ -868,15 +883,49 @@ impl<C: sov_modules_api::Context> Evm<C> {
...
+    pub fn eth_estimate_diff_size(
```
Comment (rakanalh) [inline crates/evm/src/query.rs:913]:
```suggestion
    pub fn eth_estimate_gas_delta(
```

**Review by rakanalh — COMMENTED** [review]
TBH, i am not in favor of using `diff size` as the name.

Delta is what's more commonly used to describe difference between two different values.

**Review by rakanalh — APPROVED** [review]
Thanks to @kpp for clarifying what is meant by diff here.

LGTM

**Review by eyusufatik — APPROVED** [review]

**Pull request merged**